File: JS/bintree.js
class Node {
  constructor(value) {
    this.left = null;
    this.right = null;
    this.value = value;
  }
}

class BinTree {
  constructor() {
    this.root = null;
  }
  
  insert(value) {
    const newNode = new Node(value);
    if(this.root === null) {
      this.root = newNode;
    } else {
      let currentNode = this.root;
      while(true) {
        if(value < currentNode.value){
          if(!currentNode.left) {
            currentNode.left = newNode;
            return this;
          }
          currentNode = currentNode.left;
        } else {
          if(!currentNode.right){
            currentNode.right = newNode;
            return this;
          }
          currentNode = currentNode.right;
        }
      }
    }
  }

  seek(value) {
    if(this.root === null) {
      return null;
    } else {
      let currentNode = this.root;
      while(true) {
        if(value === currentNode.value) {
          return currentNode;
        } else if (value < currentNode.value){
          currentNode = currentNode.left;
        } else {
          currentNode = currentNode.right;
        }
      }
    }

  }
}

const tree = new BinTree();
tree.insert(10);
tree.insert(4);
tree.insert(20);
tree.insert(2);
tree.insert(8);
tree.insert(17);
tree.insert(170);

File: Java/Bintree.java
package Java;

public class Bintree {
  public BTNode root;

  public Bintree() {
    this.root = null;
  }

  public Bintree insert(int value) {
    BTNode newNode = new BTNode(value);
    Bintree result = null;
    if(this.root == null) {
      this.root = newNode;
    } else {
      BTNode currentNode = this.root;
      while(true) {
        if(value < currentNode.value) {
          if(currentNode.left == null) {
            currentNode.left = newNode;
            result = this;
            break;
          }
          currentNode = currentNode.left;
        } else {
          if(currentNode.right == null) {
            currentNode.right = newNode;
            result = this;
            break;
          }
          currentNode = currentNode.right;
        }
      }
    }
    return result;
  }

  public BTNode seek(int value) {
    if(this.root == null) {
      return null;
    } else {
      BTNode currentNode = this.root;
      while(true) {
        if(value == currentNode.value) {
          return currentNode;
        } else if (value < currentNode.value) {
          currentNode = currentNode.left;
        } else {
          currentNode = currentNode.right;
        }
      }
    }
  }
}
